cli/command/stack/kubernetes/watcher.go

```diff
@@ -2,41 +2,75 @@ package kubernetes
 
 import (
 	"fmt"
+	"io"
 	"time"
 
+	composev1beta1 "github.com/docker/cli/kubernetes/client/clientset_generated/clientset/typed/compose/v1beta1"
 	apiv1beta1 "github.com/docker/cli/kubernetes/compose/v1beta1"
 	"github.com/docker/cli/kubernetes/labels"
+	"github.com/pkg/errors"
 	apiv1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
 	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
 )
 
 // DeployWatcher watches a stack deployement
-type DeployWatcher struct {
-	Pods corev1.PodInterface
+type deployWatcher struct {
+	pods   corev1.PodInterface
+	stacks composev1beta1.StackInterface
+	out    io.Writer
 }
 
 // Watch watches a stuck deployement and return a chan that will holds the state of the stack
-func (w DeployWatcher) Watch(stack *apiv1beta1.Stack, serviceNames []string) chan bool {
-	stop := make(chan bool)
+func (w *deployWatcher) Watch(stack *apiv1beta1.Stack, serviceNames []string) error {
+	err := make(chan error)
 
-	go w.waitForPods(stack.Name, serviceNames, stop)
+	go w.watchStackStatus(stack.Name, err)
+	go w.waitForPods(stack.Name, serviceNames, err)
 
-	return stop
+	return <-err
 }
 
-func (w DeployWatcher) waitForPods(stackName string, serviceNames []string, stop chan bool) {
-	starts := map[string]int32{}
+func (w *deployWatcher) watchStackStatus(stackname string, e chan error) {
+
+	watcher, err := w.stacks.Watch(metav1.ListOptions{
```

Comment:
We might prefer to use an informer (which under the hood uses a Watch and fallback to polling in case of error / connectivity loss)